### PR TITLE
Implemented Address Computer

### PR DESCRIPTION
**Pull request by popenta**
Implemented the `AddressComputer` class according to the [sdk-specs](https://github.com/multiversx/mx-sdk-specs/blob/main/core/address.md#addresscomputer).

### CODE REVIEW

File: src/smartcontracts/transactionPayloadBuilders.ts [hunk 1]
```diff
@@ -1,11 +1,10 @@
+import { WasmVirtualMachine } from "../constants";
```
Comment (andreibancioiu) [inline src/smartcontracts/transactionPayloadBuilders.ts:1]: Should be fine, we can consider that this isn't a breaking change.

File: src/address.ts [hunk 2]
```diff
@@ -259,3 +268,70 @@ export class Address {
...
 }
+
+export class AddressComputer {
+    private readonly numberOfShardsWithoutMeta: number;
+
+    constructor(numberOfShardsWithoutMeta?: number) {
+        this.numberOfShardsWithoutMeta = numberOfShardsWithoutMeta || CURRENT_NUMBER_OF_SHARDS_WITHOUT_META;
+    }
+
+    computeContractAddress(deployer: IAddress, deploymentNonce: bigint): Address {
```
Comment (andreibancioiu) [inline src/address.ts:279]: Let's call this implementation from `smartContract.ts`.

File: src/address.ts [hunk 3]
```diff
@@ -259,3 +268,70 @@ export class Address {
...
+        let initialPadding = Buffer.alloc(8, 0);
```
Comment (andreibancioiu) [inline src/address.ts:280]: `const` where possible. Here, and below.

File: src/address.ts [hunk 4]
```diff
@@ -259,3 +268,70 @@ export class Address {
...
+
+        let address = new Address(addressBytes);
+        return address;
+    }
+
+    getShardOfAddress(address: IAddress): number {
+        return this.getShardOfPubkey(address.getPublicKey(), this.numberOfShardsWithoutMeta);
+    }
+
+    private getShardOfPubkey(pubkey: Uint8Array, numberOfShards: number): number {
```
Comment (andreibancioiu) [inline src/address.ts:302]: For review, linking to the Go and PY implementations:
 - https://github.com/multiversx/mx-chain-core-go/blob/main/core/sharding/multiShardCoordinator.go#L52
 - https://github.com/multiversx/mx-sdk-py-core/blob/main/multiversx_sdk_core/address.py#L141

This works for 1, 2, 3 and 4 shards. But will fail for more. See the Go implementation. For the moment, we can keep the PY and JS implementations as they are, but we should fix them in the future.

File: src/address.spec.ts [hunk 5]
```diff
@@ -67,4 +67,33 @@ describe("test address", () => {
...
+        const addressComputer = new AddressComputer();
+        const deployer = Address.fromBech32("erd1j0hxzs7dcyxw08c4k2nv9tfcaxmqy8rj59meq505w92064x0h40qcxh3ap");
+
+        let contractAddress = addressComputer.computeContractAddress(deployer, 0n);
+        assert.equal(contractAddress.toHex(), "00000000000000000500bb652200ed1f994200ab6699462cab4b1af7b11ebd5e");
+        assert.equal(contractAddress.toBech32(), "erd1qqqqqqqqqqqqqpgqhdjjyq8dr7v5yq9tv6v5vt9tfvd00vg7h40q6779zn");
+
+        contractAddress = addressComputer.computeContractAddress(deployer, 1n);
+        assert.equal(contractAddress.toHex(), "000000000000000005006e4f90488e27342f9a46e1809452c85ee7186566bd5e");
+        assert.equal(contractAddress.toBech32(), "erd1qqqqqqqqqqqqqpgqde8eqjywyu6zlxjxuxqfg5kgtmn3setxh40qen8egy");
```
Comment (andreibancioiu) [inline src/address.spec.ts:81]: :+1: 

For reviewers: https://github.com/multiversx/mx-sdk-py-core/blob/main/multiversx_sdk_core/address_test.py

File: src/address.ts (hunk 4 above)
Comment (popenta) [reply src/address.ts:302]: Will replicate the Go implementation in the future.

File: src/smartcontracts/transactionPayloadBuilders.ts (hunk 1 above)
Comment (popenta) [reply src/smartcontracts/transactionPayloadBuilders.ts:1]: I don't see why this could be breaking.

File: src/address.ts (hunk 3 above)
Comment (popenta) [reply src/address.ts:280]: Done!

File: src/address.ts (hunk 2 above)
Comment (popenta) [reply src/address.ts:279]: Done!

File: src/smartcontracts/smartContract.ts [hunk 6]
```diff
@@ -259,26 +256,8 @@ export class SmartContract implements ISmartContract {
...
-        let addressBytes = Buffer.concat([
-            initialPadding,
-            vmTypeBytes,
-            hash.slice(10, 30),
-            shardSelector
-        ]);
-
-        let address = new Address(addressBytes);
-        return address;
+        const deployer = new Address(owner.bech32());
```
Comment (andreibancioiu) [inline src/smartcontracts/smartContract.ts:259]: Could have used the new approach, `Address.fromBech32()`.

**Review by andreibancioiu — APPROVED** [review]

File: src/smartcontracts/smartContract.ts (hunk 6 above)
Comment (popenta) [reply src/smartcontracts/smartContract.ts:259]: indeed.

File: src/smartcontracts/transactionPayloadBuilders.ts (hunk 1 above)
Comment (andreibancioiu) [reply src/smartcontracts/transactionPayloadBuilders.ts:1]: Being a publicly exported element, a movement can affect clients that use fully qualified imports (in this context, we can simply think of this as a non-breaking change, though).

**Review by andreibancioiu — APPROVED** [review]

**Review by danielailie — APPROVED** [review]

**Pull request merged**